ocaml/testsuite/tests/typing-modules/Test.ml

```diff
@@ -123,7 +123,7 @@ module F : functor (X : sig end) -> sig val x : int end
 Line 2, characters 0-3:
 2 | F.x;; (* fail *)
     ^^^
-Error: The module "F" is a functor, it cannot have any components
+Error: The module "F" is a functor, it cannot be used as a structure
```

Comment:
Now I think you are making this error worse. Would you mind making:
```ocaml
type components_reason =
  | Projection
  | Open
```
and passing it down to `lookup_structure_components` and then putting it into `No_components_abstract` and `No_components_functor`, and using it to fine tune the message.

Reply:
Done - I did it slightly differently tho, because both `Project` and `Open` are specific to structure (as opposed to functors, or generic modules), so the `reason` shouldn't go into `No_components_abstract` and `No_components_alias` (as both are for generic modules).

(I think that's in fact what you meant anyway, modulo the typo)